drivers: rcar_mmc: take into account Zephyr SDHC cmd and data timeouts

xakep-amatop opened the pull request:
Use Zephyr cmd and data timeouts where it is possible in R-car MMC driver instead of hardcoded timeouts for data and cmd xref. The default MMC driver timeout has still used in a lot of places, e.g. polling CBSY flag.

**Note: Zephyr uses CMD and data timeouts measured in ms, but in R-Car SD/MMC driver microseconds is used, so for some of variables type has been changed from `int`/`int32` to `int64` in order to deal with possible overflow in case of multiplication**

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -233,13 +233,12 @@ static int rcar_mmc_check_errors(const struct device *dev)
 static int rcar_mmc_poll_reg_flags_check_err(const struct device *dev,
 	unsigned int reg, uint32_t flag, uint32_t state, bool check_errors,
-	bool check_dma_errors)
+	bool check_dma_errors, int32_t timeout_us)
```
`timeout` field of `struct sdhc_command` has type `int`.
Removed `remain_us` had type `int`. Why this new parameter is `int32_t`? You need to propagate types from the higher level API if there are not strict reasons to not do so.

xakep-amatop replied:
int32_t and int are the same for Zephyr, take a look how zephyr defines int max:
```
#define INT_MAX		__INT_MAX__
``` 
--->
```
#if __INT_MAX__ == 2147483647
#define __INT_WIDTH__ 32
#define __SIZEOF_INT__ 4
#else
#error "unexpected __INT_MAX__ value"
#endif
```
so, here it is nothing to worry

lorc replied:
It adds cognitive load: programmer needs to wonder: "why SDHC subsystem code uses `int` and SDMMC driver - `int32_t`?". As you can see, it even started this comments thread :)

General idea of safe and secure code is that it should be obvious and boring. 

Anyways, you can leave it as is, if you wish. I am not owner of this code, so I can't force you to have it in consistent style. Just keep this in the mind for the future changes

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -259,7 +258,7 @@ static int rcar_mmc_poll_reg_flags_check_err(const struct device *dev,
 
 		k_usleep(MMC_POLL_FLAGS_ONE_CYCLE_TIMEOUT_US);
-		remain_us -= MMC_POLL_FLAGS_ONE_CYCLE_TIMEOUT_US;
+		timeout_us -= MMC_POLL_FLAGS_ONE_CYCLE_TIMEOUT_US;
```
MISRA C frowns upon changing values of function parameters (rule 17.8). I do understand that this rule is advisory and that there are plenty of places where this rule is violated. But in your case it was easier to follow it,  because resulting diff would be smaller. So, you if want, you can return back `remain_us`, just initialize it with `timeout_us`.

xakep-amatop replied:
some of calls use int32 some of calls use int literal constant
even if constant will be less then int32 (which is impossible according this comment https://github.com/xen-troops/zephyr/pull/47#discussion_r1210014393), literal constant will be promoted to higher int32 value with sign propagation

so if you don't have a crucial reason to change this code, I leave it change as is

lorc replied:
Did you replied to the right comment? My point was that you might consider not to remove `remain_us`.

xakep-amatop replied:
I see no reason to use an additional variable if we already have one which is good for the same purpose

lorc replied:
Well, strictly speaking, this is not a variable, but I got your point.

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -827,14 +827,18 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 		uint8_t *buf = (uint8_t *)data->data + (block * data->block_size);
 		uint32_t info2_reg;
 		uint16_t w_off; /* word offset in a block */
+		uint32_t elapsed_us = k_ticks_to_us_ceil32(k_uptime_ticks()) - start_data_xref_us;
```
4294967296 microseconds is only 4295 seconds, which is 1.2 hours. This means that overflow event is very likely around there.

xakep-amatop replied:
hmmm, I don't see any problem here could you please give an example with a real values and step-by-step?
 note: I see only possible problem here `data->timeout_ms * 1000`

lorc replied:
`k_ticks_to_us_ceil32(k_uptime_ticks())` might overflow between

`start_data_xref_us = k_ticks_to_us_ceil32(k_uptime_ticks());`

and

`uint32_t elapsed_us = k_ticks_to_us_ceil32(k_uptime_ticks()) - start_data_xref_us;`

This will lead to very high value in `elapsed_us`. This will lead to problems below.

xakep-amatop replied:
I don't expect that we waste 35 minutes for one step of the loop, even if it happens we will leave the loop because difference become negative

it is more interesting can we leave cycle before timeout elapsed in case if timeout less than 35 minutes?..

ok, I redo it

but there is an interesting part of code inside sd_ops:
```
int sdmmc_wait_ready(struct sd_card *card)
{
	int ret, timeout = CONFIG_SD_DATA_TIMEOUT * 1000;
```
as you can see we can get really easily an overflow during multiplication
but seems like nobody uses long timeouts...

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -827,14 +827,18 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 		uint32_t info2_reg;
 		uint16_t w_off; /* word offset in a block */
+		uint32_t elapsed_us = k_ticks_to_us_ceil32(k_uptime_ticks()) - start_data_xref_us;
+		int32_t left_time_us = data->timeout_ms * 1000 - elapsed_us;
```
`elapsed_us` here will be converted into plain `int`, which will lead to interesting results in half of the cases :)

xakep-amatop replied:
nope, see C standard about integer promotion
`elapsed_us` will have the same type in this expression, but `data->timeout_ms * 1000` will be converted to `uint32_t`

lorc replied:
Yes, and then it will be stored into `int32_t` variable. You will get large negative value there in some cases

xakep-amatop committed:
drivers: rcar_mmc: take into account Zephyr SDHC cmd and data timeouts

Use Zephyr cmd and data timeouts where it is possible in R-car MMC driver
instead of hardcoded timeouts for data and cmd xref.
The default MMC driver timeout has still used in a lot of places,
e.g. polling CBSY flag.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

lorc reviewed (approved):
`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

xakep-amatop commented:
added `reviewed-by`

firscity reviewed (approved):
In commit message - remove:

> \# Conflicts:
> \#	drivers/sdhc/rcar_mmc.c

With comment addressed - `Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

xakep-amatop commented:
> Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

done
